libraries/Microsoft.Bot.Builder.Azure/CosmosDbStorage.cs

```diff
@@ -39,13 +39,27 @@ public class CosmosDbStorage : IStorage
         /// using the provided CosmosDB credentials, database ID, and collection ID.
         /// </summary>
         /// <param name="cosmosDbStorageOptions">Cosmos DB storage configuration options.</param>
-        public CosmosDbStorage(CosmosDbStorageOptions cosmosDbStorageOptions)
+        /// <param name="jsonSerializer">Optional JsonSerializer. Mind that the TypeNameHandling, NullvalueHandling and ContractResolver will be overridden.</param>
+        public CosmosDbStorage(CosmosDbStorageOptions cosmosDbStorageOptions, JsonSerializer jsonSerializer = null)
         {
             if (cosmosDbStorageOptions == null)
             {
                 throw new ArgumentNullException(nameof(cosmosDbStorageOptions));
             }
 
+            if (jsonSerializer == null)
+            {
+                _jsonSerializer = JsonSerializer.Create(new JsonSerializerSettings { TypeNameHandling = TypeNameHandling.All });
+            }
+            else
+            {
+                // These need to be overridden to the following values.
+                jsonSerializer.TypeNameHandling = TypeNameHandling.All;
```

Comment:
I'm not clear on this - if the developer passes in a serializer, we should use it "as is". We can certainly document "We recommend the following settings", but to change what they passed in will probably lead to grief and unexpected behaviors.

Reply:
The problem is that it will crash. If these two are not sett, it will not work at all.

But yes, I might of course extract this to its own ctor. 

By the way, how its working now, did in fact lead to grief and unexpected behaviour ;-)

I'll make another "clean" version og the ctor.

Reply:
If it crashes w/o these, then we should keep them in. Adding an an XmlCode Comment on the parameter to indicate this would be enough for me.